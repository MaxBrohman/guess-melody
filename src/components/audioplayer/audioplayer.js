import React, {useRef} from 'react';

const useActiveTrack = (activeTrack, {current}) => {
  const isCurrentTrackActive = current && activeTrack === current.src;
  const className = isCurrentTrackActive
    ? `pause`
    : `play`;
  if (isCurrentTrackActive) {
    current.play();
  } else if (current) {
    current.pause();
  }
  return {
    className,
  };
};
// eslint-disable-next-line
export const Audioplayer = ({src, type, activeTrack, clickHandler}) => {
  const audioRef = useRef(null);
  const {className} = useActiveTrack(activeTrack, audioRef);
  return (
    <div className="track">
      <button
        className={`track__button track__button--${className}`}
        type="button"
        onClick={clickHandler(audioRef || {})}
      >
      </button>
      <div className="track__status">
        {// eslint-disable-next-line
          <audio src={src} ref={audioRef}></audio>}
      </div>
      {
        type === `genre` && <div className="game__answer">
          <input className="game__input visually-hidden" type="checkbox" name="answer" value={src} id={`answer-${src}`} />
          <label className="game__check" htmlFor={`answer-${src}`}>Отметить</label>
        </div>
      }
    </div>
  );
};
