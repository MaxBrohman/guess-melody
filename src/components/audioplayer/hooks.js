export const useActiveTrack = (activeTrack, {current}) => {
  const isCurrentTrackActive = current && activeTrack === current.src;
  const className = isCurrentTrackActive
    ? `pause`
    : `play`;
  if (isCurrentTrackActive) {
    current.play();
  } else if (current) {
    current.pause();
  }
  return {
    className,
  };
};
