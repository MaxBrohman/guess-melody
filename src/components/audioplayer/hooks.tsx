import {MutableRefObject} from "react";

export const useActiveTrack = (
    activeTrack: string,
    {current}: MutableRefObject<HTMLAudioElement>,
): {className: string} => {
  const isCurrentTrackActive = current && activeTrack === current.src;
  const className = isCurrentTrackActive
    ? `pause`
    : `play`;
  if (isCurrentTrackActive) {
    current.play();
  } else if (current) {
    current.pause();
  }
  return {
    className,
  };
};
