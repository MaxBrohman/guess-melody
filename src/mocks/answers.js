export const answers = [
  {
    isCorrect: true,
    isFast: false,
    time: 265,
  },
  {
    isCorrect: true,
    isFast: true,
    time: 255,
  },
  {
    isCorrect: false,
    isFast: true,
    time: 245,
  },
  {
    isCorrect: false,
    isFast: true,
    time: 235,
  },
  {
    isCorrect: true,
    isFast: false,
    time: 200,
  },
  {
    isCorrect: true,
    isFast: true,
    time: 190,
  },
  {
    isCorrect: true,
    isFast: true,
    time: 180,
  }
];
