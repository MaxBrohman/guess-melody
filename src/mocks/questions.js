export const questions = [
  {
    type: `genre`,
    answer: `blues`,
    id: `01`,
    options: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
        id: `00`,
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
        id: `01`,
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
        id: `02`
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `blues`,
        id: `03`
      }
    ]
  },
  {
    type: `artist`,
    answer: `The Stooges`,
    id: `02`,
    src: `/05 - Nine Inch Nails - Apart.mp3`,
    options: [
      {
        imgSrc: `http://placehold.it/134x134`,
        artist: `Bob Marley`
      },
      {
        imgSrc: `http://placehold.it/134x134`,
        artist: `The Stooges`
      },
      {
        imgSrc: `http://placehold.it/134x134`,
        artist: `Frank Sinatra`
      }
    ]
  },
  {
    type: `genre`,
    answer: `rock`,
    id: `03`,
    options: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
        id: `04`
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
        id: `05`
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
        id: `06`
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `rock`,
        id: `07`
      }
    ]
  }
];
