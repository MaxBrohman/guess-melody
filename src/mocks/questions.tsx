import {TQuestion} from '../reducer/types';

export const questions: Array<TQuestion> = [
  {
    type: `genre`,
    genre: `blues`,
    answers: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `blues`,
      }
    ]
  },
  {
    type: `artist`,
    song: {
      artist: `The Stooges`,
      src: `/05 - Nine Inch Nails - Apart.mp3`,
    },
    answers: [
      {
        picture: `http://placehold.it/134x134`,
        artist: `Bob Marley`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `The Stooges`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `Frank Sinatra`
      }
    ]
  },
  {
    type: `genre`,
    genre: `rock`,
    answers: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `rock`,
      }
    ]
  },
  {
    type: `artist`,
    song: {
      artist: `Frank Sinatra`,
      src: `/05 - Nine Inch Nails - Apart.mp3`,
    },
    answers: [
      {
        picture: `http://placehold.it/134x134`,
        artist: `Bob Marley`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `The Stooges`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `Frank Sinatra`
      }
    ]
  }
];
