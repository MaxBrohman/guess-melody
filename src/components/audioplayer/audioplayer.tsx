import React, {
  useRef,
  MutableRefObject,
  Fragment,
} from 'react';

import {useActiveTrack} from './hooks';

type TAudioplayerProps = {
  src: string;
  activeTrack: string;
  clickHandler: (track: MutableRefObject<HTMLAudioElement> | {}) => () => void;
}

export const Audioplayer = ({
  src,
  activeTrack,
  clickHandler,
}: TAudioplayerProps): JSX.Element => {
  const audioRef = useRef(null);
  const {className} = useActiveTrack(activeTrack, audioRef);
  return (
    <Fragment>
      <button
        className={`track__button track__button--${className}`}
        type="button"
        onClick={clickHandler(audioRef || {})}
      >
      </button>
      <div className="track__status">
        <audio src={src} ref={audioRef}></audio>
      </div>
    </Fragment>
  );
};
