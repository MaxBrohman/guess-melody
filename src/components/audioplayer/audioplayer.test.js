import React from 'react';
import renderer from 'react-test-renderer';

import {Audioplayer} from './audioplayer';
import {questions} from '../../mocks/questions';

it(`Renders Audioplayer component correctly`, () => {
  const clickHandler = jest.fn();
  const tree = renderer
                .create(
                    <Audioplayer
                      src={questions[0].options[0].src}
                      type={questions[0].type}
                      id={questions[0].options[0].id}
                      activeTrack={new Audio(questions[0].options[0].src)}
                      clickHandler={clickHandler}
                    />
                )
                .toJSON();
  expect(tree).toMatchSnapshot();
});
