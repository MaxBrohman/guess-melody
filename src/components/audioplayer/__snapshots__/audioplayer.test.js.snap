// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Audioplayer component correctly 1`] = `
<div
  className="track"
>
  <button
    className="track__button track__button--play"
    type="button"
  />
  <div
    className="track__status"
  >
    <audio
      src="/01 - Nine Inch Nails - Letting Go While Holding On.mp3"
    />
  </div>
  <div
    className="game__answer"
  >
    <input
      className="game__input visually-hidden"
      id="answer-/01 - Nine Inch Nails - Letting Go While Holding On.mp3"
      name="answer"
      type="checkbox"
      value="/01 - Nine Inch Nails - Letting Go While Holding On.mp3"
    />
    <label
      className="game__check"
      htmlFor="answer-/01 - Nine Inch Nails - Letting Go While Holding On.mp3"
    >
      Отметить
    </label>
  </div>
</div>
`;
